File: apps/android/app/src/main/java/com/douyindownload/PolicyVerifier.kt
package com.douyindownload

import android.util.Base64
import org.bouncycastle.crypto.params.Ed25519PublicKeyParameters
import org.bouncycastle.crypto.signers.Ed25519Signer
import org.json.JSONObject

/**
 * Ed25519 策略验签（防伪造停服开关）。
 *
 * 私钥仅由作者持有（scripts/sign-policy.py + secrets/policy-private-key.pem），
 * 本公钥必须与 scripts/sign-policy.py 及其他客户端保持一致。
 * 验签失败 / 缺失签名一律视为不可信策略。
 */
object PolicyVerifier {

    private const val PUBLIC_KEY_B64 = "TfI3/szbWh13QZr/FunFipeal2vb+vkrYoazGJHf6iw="

    /** 签名内容（规范化字符串，与 scripts/sign-policy.py 及各端完全一致）。 */
    fun canonicalPolicy(policy: JSONObject): String {
        val download = policy.optJSONObject("download") ?: JSONObject()
        val enabled = download.optBoolean("enabled", true)
        return listOf(
            policy.optString("updated_at", ""),
            enabled.toString(),
            download.optString("message", ""),
            download.optString("min_version", "0.0.0"),
        ).joinToString("\n")
    }

    /** 返回 true 仅当策略携带有效的 Ed25519 签名。 */
    fun verify(policy: JSONObject): Boolean {
        val signatureB64 = policy.optString("signature", "")
        if (signatureB64.isEmpty()) return false
        return try {
            val keyBytes = Base64.decode(PUBLIC_KEY_B64, Base64.DEFAULT)
            val signatureBytes = Base64.decode(signatureB64, Base64.DEFAULT)
            val payload = canonicalPolicy(policy).toByteArray(Charsets.UTF_8)

            val publicKey = Ed25519PublicKeyParameters(keyBytes, 0)
            val signer = Ed25519Signer()
            signer.init(false, publicKey) // false = verify
            signer.update(payload, 0, payload.size)
            signer.verifySignature(signatureBytes)
        } catch (e: Exception) {
            false
        }
    }
}

File: python/src/media_downloader/core/policy_verifier.py
"""Ed25519 signature verification for remote policy files.

The Android client (apps/android/.../PolicyVerifier.kt) pioneered this scheme:
the maintainer signs download-policy.json with a private key that never
leaves the release machine (scripts/sign-policy.py + secrets/), and every
client verifies the ``signature`` field against this embedded public key
before trusting the policy. Fail-closed: a missing/invalid signature means
the policy is not trusted.

Canonical payload (must stay byte-identical across ALL clients):

    updated_at
    enabled          # "true" / "false"
    message
    min_version

The public key below MUST match PolicyVerifier.kt and any other client.
"""

from __future__ import annotations

import base64
from typing import Any, Dict, Optional

# Ed25519 public key (base64, 32 bytes) -- keep in sync with
# apps/android/app/src/main/java/com/douyindownload/PolicyVerifier.kt
# and apple/Sources/MediaDownloaderCore/PolicyServices.swift
PUBLIC_KEY_B64 = "TfI3/szbWh13QZr/FunFipeal2vb+vkrYoazGJHf6iw="


def _as_str(value: Any, default: str = "") -> str:
    """Mirror org.json optString semantics: non-string values -> default."""
    return value if isinstance(value, str) else default


def canonical_policy(policy: dict) -> str:
    """Serialize the signed fields exactly like PolicyVerifier.canonicalPolicy().

    Only updated_at / enabled / message / min_version are covered by the
    signature, so adding other JSON fields (issue_url, platforms, ...) does
    not invalidate existing signatures.
    """
    download = policy.get("download")
    if not isinstance(download, dict):
        download = {}
    enabled = download.get("enabled", True)
    if isinstance(enabled, bool):
        enabled_text = "true" if enabled else "false"
    else:
        enabled_text = "true" if str(enabled).lower() == "true" else "false"
    return "\n".join([
        _as_str(policy.get("updated_at")),
        enabled_text,
        _as_str(download.get("message")),
        _as_str(download.get("min_version"), "0.0.0"),
    ])


def verify_policy_signature(policy: dict) -> bool:
    """Return True only when ``policy`` carries a valid Ed25519 signature.

    Missing signature, malformed base64 or any verification failure -> False
    (the caller must treat the policy as untrusted / unreachable).
    """
    signature = policy.get("signature") if isinstance(policy, dict) else None
    if not isinstance(signature, str) or not signature.strip():
        return False
    try:
        from nacl.exceptions import BadSignatureError
        from nacl.signing import VerifyKey

        key_bytes = base64.b64decode(PUBLIC_KEY_B64)
        signature_bytes = base64.b64decode(signature.strip())
        payload = canonical_policy(policy).encode("utf-8")
        VerifyKey(key_bytes).verify(payload, signature_bytes)
        return True
    except (BadSignatureError, ValueError, TypeError, ImportError):
        return False
